Perens stereotaxic mri

PolarBean opened the pull request:
Here I added the Perens MRI stereotaxic atlas

## Description

**What is this PR**

- [ ] Bug fix
- [ ] Addition of a new feature
- [ ] Other


## References

https://github.com/brainglobe/brainglobe-atlasapi/issues/203

## How has this PR been tested?
I ran the atlas generation script, and inspected the output. 
I attempted to run the mesh checking however I am missing the correct drivers for libGL.

adamltyson commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,402 @@
+__version__ = "3"
```
Our versioning is (currently) independent from the atlas version, so this will now make version 1.0.

```suggestion
__version__ = "0"
```

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,402 @@
+PARALLEL = True  # disable parallel mesh extraction for easier debugging
+
+
+### Additional functions #####################################################
```
To avoid code duplication, I would suggest to use existing functions to map from id to acronym and other queries on the Allen atlas. Some example relevant functionality from an IPython notebook:
```
In [11]: atlas = BrainGlobeAtlas("allen_mouse_25um")

In [12]: atlas.structures.acronym_to_id_map["mp"]
Out[12]: 673

In [13]: atlas.structures["673"]
Out[13]: {'acronym': 'mp', 'id': 673, 'name': 'mammillary peduncle', 'structure_id_path': [997, 1009, 991, 824, 46, 673], 'rgb_triplet': [204, 204, 204], 'mesh_filename': PosixPath('/home/alessandro/.brainglobe/allen_mouse_25um_v1.2/meshes/673.obj'), 'mesh': None}

In [14]: atlas.structures["673"]["acronym"]
Out[14]: 'mp'

In [15]: atlas.get_structure_ancestors("673")
Out[15]: ['root', 'fiber tracts', 'mfbs', 'mfsbshy', 'mfbsma']
```

PolarBean replied:
I tried to address this in the latest commit, let me know if its what you had in mind.

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,402 @@
+            "Firefox/129.0"
+        ),
+        "Accept": (
+            "text/html,"
+            "application/xhtml+xml,"
+            "application/xml;q=0.9,"
+            "image/avif,"
+            "image/webp,"
+            "image/png,"
+            "image/svg+xml,"
+            "*/*;q=0.8"
+        ),
+        "Accept-Language": "en-GB,en;q=0.5",
+        "Accept-Encoding": "gzip, deflate, br, zstd",
+        "DNT": "1",
+        "Sec-GPC": "1",
+        "Host": "www.neuropedia.dk",
+        "Connection": "keep-alive",
+        "Upgrade-Insecure-Requests": "1",
+        "Sec-Fetch-Dest": "document",
+        "Sec-Fetch-Mode": "navigate",
+        "Sec-Fetch-Site": "none",
+        "Sec-Fetch-User": "?1",
+        "TE": "trailers",
+        "Priority": "u=0, i",
+    }
+    if not os.path.isdir(
+        atlas_files_dir / "Multimodal_mouse_brain_atlas_files"
+    ):
+        req = urllib.request.Request(ATLAS_FILE_URL, headers=headers)
+        with (
+            urllib.request.urlopen(req) as response,
+            open(destination_path, "wb") as out_file,
+        ):
+            data = response.read()  # a `bytes` object
+            out_file.write(data)
+        with py7zr.SevenZipFile(destination_path, mode="r") as z:
+            z.extractall(path=atlas_files_dir)
+        destination_path.unlink()
+
```
`pooch` _should_ be able to take care of all of the below, and would make this considerably shorter, and more consistent with newer atlas packaging scripts (we aim to move all atlas packaging to `pooch` one day #284 ). See [`axolotl_atlas.py` for an example](https://github.com/brainglobe/brainglobe-atlasapi/blob/086ef0db7761ad3051d544d0228aa448f03246eb/brainglobe_atlasapi/atlas_generation/atlas_scripts/axolotl_atlas.py#L57C2-L64C10).

PolarBean replied:
Unfortunately it seems pooch cannot handle this as it requires specific headers (The gubra server has cloudflare protection which refuses connections that are not from a browser. so my workaround is to set the request headers to that of a browser). I tried to implement a similar solution in pooch with no luck.

alessandrofelder replied:
Great detective work, thanks - apologies for leading you down that rabbit hole!
I'd suggest encapsulating this part into a function, and document why we need this instead in its docstring?

PolarBean replied:
Great idea, done

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,402 @@
+        / "mri_temp.nii.gz"
+    )
+
+    annotated_volume = sitk.GetArrayFromImage(
```
For consistency with other packaging scripts, can we use `brainglobe_utils.IO.image.load_any` instead of ITK here and lines below?

PolarBean replied:
done

alessandrofelder commented on pyproject.toml:
```diff
@@ -56,6 +56,7 @@ allenmouse = ["allensdk"]
 allenmouse_barrels = [
     "allensdk",
     "voxcell"]
+perens_stereotaxic_mri_mouse = ["py7zr"]
```
hopefully we don't need an extra dependency here, if we move to `pooch`?

PolarBean replied:
I think we still do. I will move the script to pooch but this file is in 7zip format which I think requires the py7zr dependency to decompress.

alessandrofelder replied:
You're right - yep, thank yoU!

alessandrofelder reviewed (commented):
Thanks a lot for this @PolarBean 
I've left some high-level comments for now - let me know what you think.

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,325 @@
+### Additional functions #####################################################
+
+
+def tree_traverse_child2parent(df, child_id, ids):
+    parent = df["parent_id"][df["id"] == child_id].item()
+
+    if not np.isnan(parent):
+        id = df["id"][df["id"] == parent].item()
+        ids.append(id)
+        tree_traverse_child2parent(df, parent, ids)
+        return ids
+    else:
+        return ids
```
I think this function is now unused (and should therefore be removed)?

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,325 @@
+from brainglobe_atlasapi.atlas_generation.wrapup import wrapup_atlas_from_data
+from brainglobe_atlasapi.structure_tree_util import get_structures_tree
+
+atlas = BrainGlobeAtlas("allen_mouse_25um")
```
I would maybe rename this and add a comment for legibility? So we avoid confusion between the atlas packaged here and the Allen one? 
```suggestion
# The Perens atlas re-uses information from the Allen atlas, so it's useful to have an instance of the Allen atlas around
allen_atlas = BrainGlobeAtlas("allen_mouse_25um")
```

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,325 @@
+
+
+def download_and_extract_files(ATLAS_FILE_URL, destination_path):
+    """This is needed to get the brainglobe data from their server,
```
```suggestion
    """This is needed to get the source data from their server,
```

PolarBean replied:
oops

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,325 @@
+    Returns:
+        parents (list) : brain region acronym corresponding to input ID
+    """
+    parents = atlas.get_structure_ancestors(temp_id)
```
```suggestion
parents = atlas.get_structure_ancestors(key)
```
I think?

Also, not sure these two lines are worth an additional function? It's only used once, and could just be an extra line in the `for` loop to reduce levels of abstraction?

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,325 @@
+    Get all parent IDs/acronyms in Allen's brain atlas hierarchical structure'
+
+    Call:
+        get_all_children(df, key)
```
```suggestion
        get_all_parents(df, key)
```

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,325 @@
+
+def get_all_parents(atlas, key):
+    """
+    Get all parent IDs/acronyms in Allen's brain atlas hierarchical structure'
```
this function is not specific to the Allen, right? But as I mention below, I'd be tempted to remove this function altogether?

PolarBean replied:
I've instead used the bgapi to simply replicate the allen atlas structure list since they are the same. I can change this if it's not recommended.

alessandrofelder reviewed (changes requested):
Looking good - thanks @PolarBean 
Some more very minor comments.

adamltyson commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,336 @@
+        ):
+            data = response.read()  # a `bytes` object
+            out_file.write(data)
+        with py7zr.SevenZipFile(destination_path, mode="r") as z:
```
Is there a way to do this without adding a new dependency (`py7zr`)?

adamltyson replied:
Cool, just wondering.

adamltyson commented:
@PolarBean is it possible to add a few more print statements, and ideally a download progress bar? It's hard to know what's going on when running this script.

adamltyson commented:
Thanks for working on this @PolarBean. 

I'm not exactly sure what files are packaged with this atlas, but presumably we should be packaging up multiple reference (template) images? If I understand correctly, the main reason to add this atlas would be to make available all the different reference images as registration targets? 

It may also be worth packaging up multiple atlases, one for each coordinate space.

PolarBean commented:
> Thanks for working on this @PolarBean.
> 
> I'm not exactly sure what files are packaged with this atlas, but presumably we should be packaging up multiple reference (template) images? If I understand correctly, the main reason to add this atlas would be to make available all the different reference images as registration targets?
> 
> It may also be worth packaging up multiple atlases, one for each coordinate space.

This atlas only has one template image (MRI). It is in a different space to the perens LSFM atlas. It's valuable because it is in flat skull position and is therefore useful for stereotaxic surgery. It is also useful as a MRI registration target. I will make a seperate pull request for the most recent perens lsfm volume once they have updated those files on their server.

adamltyson commented:
Ah, nvm. I assumed as it was a "multimodal" atlas framework, there were templates for LSFM & CT in the same space.

PolarBean commented:
Yeah they have multiple templates in multiple spaces and they are linked by deformation matrices 
I've also developed https://github.com/Neural-Systems-at-UIO/CCF_translator to make transferring data between the spaces more user friendly.

PolarBean commented:
> If I understand correctly, the main reason to add this atlas would be to make available all the different reference images as registration targets?
Also one thing I'm curious about is what should our approach be to adding new resources to brainglobe. Should a new atlas provide some benefit in order to be added, or instead is just the existence of a resource enough to add it (even if it has been superseded by newer better resources)? I would tend towards wanting to add in everything, as it could be useful for say replicating past results or comparing atlas definitions of particular regions. But others may disagree.

adamltyson commented:
I agree with you. As an example [this atlas](https://brainglobe.info/documentation/brainglobe-atlasapi/usage/atlas-details.html#smoothed-version-of-the-kim-et-al-mouse-reference-atlas) was added to allow for some pre-BrainGlobe analyses to be replicated with BrainGlobe tools.

If an atlas has obviously been replaced by another, then I would suggest we:
- Make sure this is noted on the website to ensure users are using the best atlas for them
- We don't prioritise adding it, and try to focus on the atlases with the biggest impact (which I assume will be the "better" ones)

PolarBean commented:
> @PolarBean is it possible to add a few more print statements, and ideally a download progress bar? It's hard to know what's going on when running this script.

No worries, I've added a download progress bar and a few more print statements. Let me know if it's what you had in mind

adamltyson commented:
> No worries, I've added a download progress bar and a few more print statements. Let me know if it's what you had in mind

Yes, thanks! A bit more printing (e..g just to say what's being downloaded might be useful too).

I haven't checked the packaged atlas in detail (I'll leave that to @alessandrofelder when he's back), but I noticed a couple of issues:
- The packaged atlas is in `ars` orientation (rather than `asr`). This would suggest that the orientation set in the script is wrong. This may be because `load_any()` is loading the raw image data, and not transforming the data according to the nifti metadata
- There are many more stuctures (840) than meshes (754). It's not always possible to create a mesh for every region, but nearly 100 missing is surprising. Not sure why this has happened though.

PolarBean commented:
> * The packaged atlas is in `ars` orientation (rather than `asr`). This would suggest that the orientation set in the script is wrong. This may be because `load_any()` is loading the raw image data, and not transforming the data according to the nifti metadata

I see, I'll have to recheck it. I thought it was unchanged from the previous perens lsfm in the apiwhich is listed as IAL and so I copied it from there. I will recheck it for this one and for the other pull requests I have.

PolarBean commented:
>     * There are many more stuctures (840) than meshes (754). It's not always possible to create a mesh for every region, but nearly 100 missing is surprising. Not sure why this has happened though.

my guess is that its to do with this atlas being derived from the allen but is lower resolution. Therefore many of the meshes which are large enough to be rendered in the allen atlas (particularly in 10um) are too small here. 

Its actually surprising that this atlas has a voxel size of 25um but contains fewer voxels than does the allen 25um. I had thought that LSFM actually expanded the brain, not shrank it. but I may be confusing things with expansion microscopy.

PolarBean commented:
> * There are many more stuctures (840) than meshes (754). It's not always possible to create a mesh for every region, but nearly 100 missing is surprising. Not sure why this has happened though.

Hmm, I contacted Gubra about some problems I was facing when ingesting the LSFM volume (for which I will soon open a pull request). They fixed that problem and it seems they also updated the MRI annotation while they were at it. When I run this script I now get 838 meshes which seems much more respectable!

PolarBean commented:
I read the documentation but Im a bit confused about differentiating between ars and als for symmetrical atlases. I put als as i thought left to right seemed most intuitive but I am a bit lost.

adamltyson commented:
> I read the documentation but Im a bit confused about differentiating between ars and als for symmetrical atlases. I put als as i thought left to right seemed most intuitive but I am a bit lost.

In principle that shouldn't make a difference.

PolarBean commented:
> A bit more printing (e..g just to say what's being downloaded might be useful too).

it actually only downloads a single file as all the atlas data is distributed in a single archive. I'll add print statements elsewhere though

PolarBean commented:
ok actually i added one more print statement but it seems now every function has at least one print statement. I think this I've addressed everything now but feel free to add more reccomendations. I'll also update the template script with print statements so it should be included by default in the future.

PolarBean commented:
I ran the validation script Alessandro sent me and it returned the following error
```
Failed validations:
('catch_missing_mesh_files', "Structures with IDs [545, 1051] are in the atlas, but don't have a corresponding mesh file.")
WARNING: Could not load the Qt platform plugin "xcb" in "" even though it was found.
WARNING: This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, minimal, minimalegl, offscreen, vnc, webgl, xcb.

Aborted (core dumped)
```

I'll try to fix the xcb error but the missing structures are likely just due to those regions being too small in this atlas.

PolarBean commented:
![image](https://github.com/user-attachments/assets/af89264b-ea2f-4cf9-be91-a3602100ed26)
ok after resolving the difficulties the validation script ran and showed me this volume. It all looks as expected!

alessandrofelder commented:
Thanks @PolarBean - I'll review next week!

PolarBean commented:
great! how'd you go?

alessandrofelder commented on brainglobe_atlasapi/atlas_generation/atlas_scripts/perens_stereotaxic_mri_mouse.py:
```diff
@@ -0,0 +1,344 @@
+ATLAS_FILE_URL = "https://www.neuropedia.dk/wp-content/uploads/Multimodal_mouse_brain_atlas_files_v2.7z"
+# The orientation of the **original** atlas data, in BrainGlobe convention:
+# https://brainglobe.info/documentation/setting-up/image-definition.html#orientation
+ORIENTATION = "als"
```
```suggestion
ORIENTATION = "lai"
```

alessandrofelder reviewed (commented):
Almost there.

Looks like I still get two regions without a mesh in the validation:
>('catch_missing_mesh_files', "Structures with IDs [545, 1051] are in the atlas, but don't have a corresponding mesh file.")

But my guess is those regions are so small (1051 is one pixel in the annotation, I think) that our code didn't mesh them "on purpose".

The atlas doesn't come out oriented in BrainGlobe ASR for me though, maybe [because of our `load_any` function ignoring nifti metadata](https://github.com/brainglobe/brainglobe-atlasapi/pull/370#issuecomment-2346746480). I think the original data is LAI in brainglobe terms (from looking at the raw MRI data in napari, loaded with `load_any`. 

I'll run again with my one-line suggestion to check this makes the BG version ASR and report back here.

alessandrofelder commented on README.md:
```diff
@@ -41,6 +41,7 @@ A number of atlases are in development, but those available currently are:
 | [BlueBrain Barrel Cortex Atlas](https://doi.org/10.1162/imag_a_00209) | 10 and 25 micron | P56 | STPT | allen_mouse_bluebrain_barrels_10um, allen_mouse_bluebrain_barrels_25um
 | [UNAM Axolotl Brain Atlas](https://doi.org/10.1038/s41598-021-89357-3) | 40 micron | ~ 3 months post hatching | MRI | unam_axolotl_40um
 | [Prairie Vole Brain Atlas](https://doi.org/10.7554/eLife.87029.3.sa0) | 25 micron | Unknown | LSFM | prairie_vole_25um
+| [Gubra's stereotaxic MRI mouse brain atlas](https://doi.org/10.1007/s12021-023-09623-9) | 25 micron | 10-week-old  | MRI
```
```suggestion
| [Gubra's stereotaxic MRI mouse brain atlas](https://doi.org/10.1007/s12021-023-09623-9) | 25 micron | 10-week-old  | MRI | perens_stereotaxic_mouse_mri_25um
```

alessandrofelder reviewed (approved):
Atlas is on GIN - thanks for your patience @PolarBean
